ComfortZoneApp/app/src/main/java/com/example/comfortzone/utils/WeatherDbUtil.java

```diff
@@ -1,68 +1,83 @@
-                    Coordinates coordinates = new WeatherData.Coordinates(lat, lon);
-                    getAndStoreWeatherData(context, coordinates);
+            public void call(Subscriber<? super Object> subscriber) {
+                if (db.weatherDao().getAll().isEmpty()) {
+                    client.getCityData(new CityListCallback() {
+                        @Override
+                        public void onGetCityList(WeatherData[] weatherData) {
+                            db.weatherDao().insertAll(weatherData);
+                            saveNewTemps(context).subscribe(subscriber);
+                        }
+                    });
+                } else {
+                    saveNewTemps(context).subscribe(subscriber);
                 }
             }
         });
 
+    }
 
-
+    private static Observable<Object> saveNewTemps(Activity activity) {
+        return Observable.create(new Observable.OnSubscribe<Object>() {
+            @Override
+            public void call(Subscriber<? super Object> subscriber) {
+                LocationUtil.getLastLocation(activity, new LocationCallback() {
+                    @Override
+                    public void onLocationUpdated(String lat, String lon) {
+                        Long timeNow = System.currentTimeMillis();
+                        Long hourAgo = timeNow - TimeUnit.HOURS.toMillis(1L);
+                        if (db.weatherDao().isUploadedOverHrAgo(hourAgo)) {
+                            Coordinates coordinates = new WeatherData.Coordinates(lat, lon);
+                            Observable<Object> ob = getAndStoreWeatherData(activity, coordinates); // returns one observable
+                            ob.subscribe(subscriber);
+                        } else {
+                            subscriber.onCompleted();
+                        }
+                    }
+                });
+            }
+        });
     }
 
-    private static void getAndStoreWeatherData(Activity context, Coordinates coordinates) {
+    private static Observable<Object> getAndStoreWeatherData(Activity context, Coordinates coordinates) {
         WeatherClient client = new WeatherClient();
         List<String> groupUrls = client.getGroupWeatherUrl(context);
+        List<Observable<Object>> listObservables = new ArrayList<>();
         for (String apiUrl : groupUrls) {
-            client.getGroupWeatherData(apiUrl, new WeatherCallback() {
-                @Override
-                public void onGetWeatherData(String data) {
-                    Gson gson = new Gson();
-                    WeatherGroupData weathers = gson.fromJson(data, WeatherGroupData.class);
-                    saveCities(weathers, coordinates);
-                }
-            });
+            listObservables.add(client.getGroupWeatherData(apiUrl, coordinates));
         }
+        Observable<Object> mergedObservables = Observable.merge(listObservables);
+        return mergedObservables;
```

Comment:
nice! since you are returning right away, we don't need to assign to a variable

i.e. `return Observable.merge(listObservables);`